src/stdlib/interpreterLazyList.ts

```diff
@@ -0,0 +1,177 @@
+import { stringify } from '../utils/stringify'
+import { isInterpreterThunk } from '../interpreter/lazyInterpreter'
+import { force, force_once } from './interpreterLazyTypeCheck'
+
+// list.ts: Supporting lists in the Scheme style, using pairs made
+//          up of two-element JavaScript array (vector)
+// Author: Martin Henz
+// Translated to TypeScript by Evan Sebastian
+export type Pair<H, T> = [H, T]
+export type List = null | NonEmptyList
+interface NonEmptyList extends Pair<any, any> {}
+
+// array test works differently for Rhino and
+// the Firefox environment (especially Web Console)
+function array_test(x: any) {
+  if (isInterpreterThunk(x)) {
+    const result = force_once(x)
+    if (Array.isArray === undefined) {
+      return result instanceof Array
+    } else {
+      return Array.isArray(result)
+    }
+  } else {
+    if (Array.isArray === undefined) {
+      return x instanceof Array
+    } else {
+      return Array.isArray(x)
+    }
+  }
+}
+
+// pair constructs a pair using a two-element array
+// LOW-LEVEL FUNCTION, NOT SOURCE
+export function pair<H, T>(x: H, xs: T): Pair<H, T> {
+  return [x, xs]
+}
+
+// is_pair returns true iff arg is a two-element array
+// LOW-LEVEL FUNCTION, NOT SOURCE
+export function is_pair(x: any) {
+  if (isInterpreterThunk(x)) {
+    const result = force_once(x)
+    return array_test(result) && result.length === 2
+  } else {
+    return array_test(x) && x.length === 2
+  }
+}
+
+// head returns the first component of the given pair,
+// throws an exception if the argument is not a pair
+// LOW-LEVEL FUNCTION, NOT SOURCE
+export function head(xs: any) {
+  if (isInterpreterThunk(xs)) {
+    const result = force(xs)
+    if (result !== null && is_pair(result)) {
+      // If head is a thunk, force its value.
+      if (result[0] !== null && isInterpreterThunk(result[0])) {
```

Comment:
It's also possible to delay the evaluation of the head as much as possible, until it's needed.

Reply:
I agree.

Reply:
You are right that it can be delayed. The reason it was made to behave this way is because for this lazy interpreter, function call statements will return the final value. For example, if we have a function `add(x, y)` that returns `x + y`, then the statement `add(1, 2);` will return `3` and not a thunk.

Having said that, there is a way to delay its evaluation by assigning it to a constant.
`const a = pair(1, 2);`
`const b = head(a);` - This will not evaluate the call to head until the value of b is needed.

However,
`head(a);` will force it to evaluate and will return `1`

Reply:
ok

Reply:
Background info on laziness of pair, head and tail: 
https://sicp.comp.nus.edu.sg/chapters/84#p4 :
Note that these lazy lists are even lazier than the streams of chapter 3: The head of the list, as well as the tail, is delayed.[2] In fact, even accessing the head or tail of a lazy pair need not force the value of a list element. The value will be forced only when it is really needed—e.g., for use as the argument of a primitive, or to be printed as an answer.